File: client/screens/game/scripts/components/creature/PlayerController.cs
using System;
using System.Reactive.Linq;
using System.Reactive.Subjects;
using Client.scripts.extensions;
using Client.scripts.global.udp.egress;
using Godot;

namespace Client.screens.game.scripts.components.creature
{
    public class PlayerController : Spatial
    {
        private bool _isRequestingPositionChange;
        private Camera _camera;
        private Subject<bool> _requestPositionChange = new();
        private Subject<bool> _unsubscribe = new();

        public override void _Ready()
        {
            GD.Print("Hello from Main Player Controller C#");
            _camera = GetNode<Camera>("Camera");
            _requestPositionChange
                .ThrottleTime(TimeSpan.FromMilliseconds(200))
                .TakeUntil(_unsubscribe)
                .Subscribe(RxObserverAdapter<bool>.Get(_ => SendPositionChange()));
        }
        protected override void Dispose(bool disposing)
        {
            base.Dispose(disposing);
            _unsubscribe.OnNext(true);
        }

        public override void _Input(InputEvent @event)
        {
            if (@event is InputEventMouseButton eventMouseButton)
            {
                if (eventMouseButton.ButtonIndex == (int) ButtonList.Right)
                {
                    if (eventMouseButton.Pressed)
                    {
                        _isRequestingPositionChange = true;
                        RequestPositionUpdate();
                    }
                    else
                    {
                        _isRequestingPositionChange = false;
                    }
                }
            }
        }

        public override void _Process(float delta)
        {
            if (_isRequestingPositionChange)
            {
                RequestPositionUpdate();
            }
        }

        private void RequestPositionUpdate()
        {
            _requestPositionChange.OnNext(true);
        }

        private void SendPositionChange()
        {
            var rayLength = 1000;
            var mousePos = GetViewport().GetMousePosition();
            var from = _camera.ProjectRayOrigin(mousePos);
            var to = from + _camera.ProjectRayNormal(mousePos) * rayLength;
            var spaceState = GetWorld().DirectSpaceState;
            var result = spaceState.IntersectRay(from, to, collideWithAreas: true, collideWithBodies: false);
            if (result.Contains("position"))
            {
                var position = (Vector3) result["position"];
                ActionSenderMono.Instance.Send(new EgressDataPacket.PositionChange(position.to2D()));
            }
        }
    }
}

File: client-jvm/src/main/kotlin/clientjvm/scenes/game/scenes/gameviewport/scenes/creatures/scenes/creature/PlayerController.kt
package clientjvm.scenes.game.scenes.gameviewport.scenes.creatures.scenes.creature

import clientjvm.exts.castCameraRays
import clientjvm.exts.convert
import clientjvm.exts.emitter
import clientjvm.exts.to2D
import clientjvm.global.ClientDataSender
import godot.GlobalConstants
import godot.InputEvent
import godot.InputEventMouseButton
import godot.Spatial
import godot.annotation.RegisterClass
import godot.annotation.RegisterFunction
import io.reactivex.rxjava3.subjects.PublishSubject
import pl.humberd.udp.packets.clientserver.PositionChange
import java.util.concurrent.TimeUnit

@RegisterClass
class PlayerController : Spatial() {
    private var mousePressed = false
    private val positionChangeStream = PublishSubject.create<Boolean>()

    private val unsub by emitter()

    @RegisterFunction
    override fun _ready() {
        positionChangeStream
            .throttleLast(200, TimeUnit.MILLISECONDS)
            .takeUntil(unsub)
            .subscribe { sendPositionChange() }
    }

    @RegisterFunction
    override fun _process(delta: Double) {
        if (mousePressed) {
            positionChangeStream.onNext(true)
        }
    }

    @RegisterFunction
    override fun _input(event: InputEvent) {
        if (event is InputEventMouseButton) {
            if (event.buttonIndex == GlobalConstants.BUTTON_RIGHT) {
                // fixme: should be event.pressed
                mousePressed = event.doubleclick
            }
        }
    }

    override fun _onDestroy() {
        unsub.onNext(true)
    }

    private fun sendPositionChange() {
        val result = castCameraRays(collideWithAreas = true, collideWithBodies = false)

        if (result != null) {
            ClientDataSender.send(PositionChange(result.position.to2D().convert()))
        }
    }
}
